src/trees/prufer.jl

```diff
@@ -0,0 +1,91 @@
+"""
+    is_tree(g)
+
+Returns true if g is a tree: that is, a simple, connected undirected graph, with nv-1 edges (nv = number of vertices). Trees are the minimal connected graphs; equivalently they have no cycles.  
+
+This function does not apply to directed graphs. Directed trees are sometimes called [polytrees](https://en.wikipedia.org/wiki/Polytree)). 
+
+"""
+
+function is_tree(g::AbstractGraph)
+    is_directed(g) && throw(ArgumentError("g must be undirected. "))
+    return is_connected(g) && ne(g) == nv(g) - 1
+end
+
+function _is_prufer(c)
+    return ndims(c) == 1 && !isempty(c) && maximum(c) <= length(c) + 2
+end
+
+function _degree_from_prufer(c::Vector{T})::Vector{T} where {T<:Integer}
+    """
+    Degree sequence from prufer code.
+    Returns d such that d[i] = 1 + number of occurences of i in c
+    """
+    n = length(c) + 2
+    return [T(count(==(i), c) + 1) for i in 1:n]
+end
+
+"""
+    prufer_decode(code)
+
+Returns the unique tree associated with the given (Prüfer) code. 
+Each tree of size n is associated with a Prüfer sequence (a[1], ..., a[n-2]) with 1 ⩽ a[i] ⩽ n. The sequence is constructed recursively by the leaf removal algoritm. At step k, the leaf with smallest index is removed and its unique neighbor is added to the Prüfer sequence, giving a[k]. The decoding algorithm goes backward.  
+Ref: [Prüfer sequence on Wikipedia](https://en.wikipedia.org/wiki/Pr%C3%BCfer_sequence)
+"""
+
+function prufer_decode(code::Array{T,1})::Graph{T} where {T<:Integer}
```

Comment:
```suggestion
function prufer_decode(code::AbstractVector{T})::SimpleGraph{T} where {T<:Integer}
```
Probably more clear what is happening, if we use `SimpleGraph` here, `Graph` is just an alias for that. It is also not that usual to annotate return types in Julia - if I remember correctly it will call `convert(SimpleGraph{T}, g)` on the result `g` of that function - this will probably just be NOP instruction here, so I don't mind.

Reply:
`Vector{T}` is an alias for `Array{T, 1}` and `AbstractVector{T}` is an alias for `AbstractArray{T, 1}` - so in order to allow other types of vectors, we can maybe use `AbstractVector{T}` here.

Reply:
I would also debate, if the return type should be inferred from the type of the array, or if we should allow one to give an optional argument specifying the resulting graph eltype. See here for some for something similar: https://github.com/JuliaGraphs/Graphs.jl/pull/197

Reply:
Lastly, maybe we should move functions that only work with `SimpleGraph` and `SimpleDiGraph` to the `SimpleGraphs` submodule?

Reply:
Regarding return types: since there is an `encode` and `decode` function, I wanted to impose the return type directly from the input, in order to guarantee `decode(encode(g))==g`. If the return type was an argument this would not be granted directly.

Maybe you suggest to have two methods `encode(T::Type,g::SimpleGraph)` and `encode(g::SimpleGraph)`, and in the second method the type is directly inferred from `g` as in my function ? I'm not sure this would be really useful...